src/EditorFeatures/Next/Options/EditorConfigDocumentOptionsProvider.cs

```diff
@@ -0,0 +1,137 @@
+
+            workspace.DocumentOpened += Workspace_DocumentOpened;
+            workspace.DocumentClosed += Workspace_DocumentClosed;
+        }
+
+        private void Workspace_DocumentClosed(object sender, DocumentEventArgs e)
+        {
+            lock (_gate)
+            {
+                Task<ICodingConventionContext> contextTask;
+
+                if (_openDocumentContexts.TryGetValue(e.Document.Id, out contextTask))
+                {
+                    _openDocumentContexts.Remove(e.Document.Id);
+
+                    // Ensure we dispose the context, which we'll do asynchronously
+                    contextTask.ContinueWith(t => t.Result.Dispose(), CancellationToken.None, TaskContinuationOptions.OnlyOnRanToCompletion, TaskScheduler.Default);
+                }
+            }
+        }
+
+        private void Workspace_DocumentOpened(object sender, DocumentEventArgs e)
+        {
+            lock (_gate)
+            {
+                _openDocumentContexts.Add(e.Document.Id, Task.Run(() => _codingConventionsManager.GetConventionContextAsync(e.Document.FilePath, CancellationToken.None)));
+            }
+        }
+
+        public async Task<IDocumentOptions> GetOptionsForDocumentAsync(Document document, CancellationToken cancellationToken)
+        {
+            Task<ICodingConventionContext> contextTask;
+
+            lock (_gate)
+            {
+                _openDocumentContexts.TryGetValue(document.Id, out contextTask);
+            }
+
+            if (contextTask != null)
+            {
+                // The file is open, let's reuse our cached data for that file. The task might still be running, but we want to allow for eager cancellation
+                // if the caller doesn't need it.
+                var cancellableContextTask = contextTask.ContinueWith(
+                    t => t.Result,
+                    cancellationToken,
+                    TaskContinuationOptions.OnlyOnRanToCompletion | TaskContinuationOptions.ExecuteSynchronously,
+                    TaskScheduler.Default);
+
+                var context = await cancellableContextTask.ConfigureAwait(false);
+                return new DocumentOptions(context.CurrentConventions);
+            }
+            else
+            {
+                var path = document.FilePath;
+
+                // The file might not actually have a path yet, if it's a file being proposed by a code action. We'll guess a file path to use
+                if (path == null)
+                {
+                    if (document.Name != null && document.Project.FilePath != null)
+                    {
+                        path = Path.Combine(Path.GetDirectoryName(document.Project.FilePath), document.Name);
+                    }
+                    else
+                    {
+                        // Really no idea where this is going, so bail
+                        return null;
```

Comment:
have we doc'ed the interface to make it clear this is a possibility?  I would not have expected it.